File: app/src/main/java/app/kamy/saatApp/domain/faraidh/FaraidhSilsilahBuilder.kt
package app.kamy.saatApp.domain.faraidh

import java.math.BigDecimal
import java.math.RoundingMode

object FaraidhSilsilahBuilder {

    private const val DECEASED_ID = "deceased"

    fun build(
        profile: DeceasedProfile,
        input: HeirInput,
        activeShares: List<HeirShare>,
        blocked: List<BlockedHeir>,
        names: FaraidhParticipantNames = FaraidhParticipantNames()
    ): List<SilsilahNode> {
        val nodes = mutableListOf<SilsilahNode>()
        val activeTypes = activeShares.map { it.type }.toSet()
        val blockedMap = blocked.groupBy { it.type }

        val deceasedLabel = profile.name.ifBlank { "" }
        nodes += SilsilahNode(
            id = DECEASED_ID,
            parentNodeId = null,
            generationLevel = 0,
            type = HeirType.SON,
            labelKey = "faraidh_node_deceased",
            headCount = 1,
            inherits = false,
            blocked = false,
            blockReason = null,
            displayName = deceasedLabel
        )

        if (input.fatherCount > 0) {
            val reason = blockedMap[HeirType.FATHER]?.firstOrNull()?.reason
            nodes += node(
                "father", DECEASED_ID, -1, HeirType.FATHER, "faraidh_heir_father", 1,
                activeTypes.contains(HeirType.FATHER), reason != null, reason, names.fatherName, activeShares, profile.bornOutOfWedlock
            )
        }
        if (input.motherCount > 0) {
            val reason = blockedMap[HeirType.MOTHER]?.firstOrNull()?.reason
            nodes += node(
                "mother", DECEASED_ID, -1, HeirType.MOTHER, "faraidh_heir_mother", 1,
                activeTypes.contains(HeirType.MOTHER), reason != null, reason, names.motherName, activeShares, profile.bornOutOfWedlock
            )
        }

        if (profile.gender == DeceasedGender.MALE && input.wifeCount > 0) {
            repeat(input.wifeCount) { i ->
                val wifeName = names.wifeNames.getOrNull(i).orEmpty()
                val reason = blockedMap[HeirType.WIFE]?.firstOrNull()?.reason
                nodes += node(
                    "wife_$i", DECEASED_ID, 0, HeirType.WIFE, "faraidh_heir_wife",
                    1, activeTypes.contains(HeirType.WIFE), reason != null, reason, wifeName, activeShares, profile.bornOutOfWedlock
                )
            }
        }
        if (profile.gender == DeceasedGender.FEMALE && input.husbandCount > 0) {
            val reason = blockedMap[HeirType.HUSBAND]?.firstOrNull()?.reason
            nodes += node(
                "husband", DECEASED_ID, 0, HeirType.HUSBAND, "faraidh_heir_husband",
                1, activeTypes.contains(HeirType.HUSBAND), reason != null, reason, names.husbandName, activeShares, profile.bornOutOfWedlock
            )
        }

        addChildBranch(nodes, input, activeTypes, blockedMap, names, activeShares, profile.bornOutOfWedlock)
        addSiblingBranch(nodes, input, activeTypes, blockedMap, names, activeShares, profile.bornOutOfWedlock)

        return nodes
    }

    private fun addChildBranch(
        nodes: MutableList<SilsilahNode>,
        input: HeirInput,
        activeTypes: Set<HeirType>,
        blockedMap: Map<HeirType, List<BlockedHeir>>,
        names: FaraidhParticipantNames,
        activeShares: List<HeirShare>,
        bornOutOfWedlock: Boolean
    ) {
        if (input.sonCount > 0) {
            repeat(input.sonCount) { i ->
                val reason = blockedMap[HeirType.SON]?.firstOrNull()?.reason
                nodes += node(
                    "son_$i", DECEASED_ID, 1, HeirType.SON, "faraidh_heir_son", 1,
                    activeTypes.contains(HeirType.SON), reason != null, reason,
                    names.sonNames.getOrNull(i).orEmpty(), activeShares, bornOutOfWedlock
                )
            }
        }
        if (input.daughterCount > 0) {
            repeat(input.daughterCount) { i ->
                val reason = blockedMap[HeirType.DAUGHTER]?.firstOrNull()?.reason
                nodes += node(
                    "daughter_$i", DECEASED_ID, 1, HeirType.DAUGHTER, "faraidh_heir_daughter", 1,
                    activeTypes.contains(HeirType.DAUGHTER), reason != null, reason,
                    names.daughterNames.getOrNull(i).orEmpty(), activeShares, bornOutOfWedlock
                )
            }
        }
        if (input.grandsonCount > 0) {
            val reason = blockedMap[HeirType.GRANDSON]?.firstOrNull()?.reason
            repeat(input.grandsonCount) { i ->
                nodes += node(
                    "grandson_$i", "son_0", 2, HeirType.GRANDSON, "faraidh_heir_grandson", 1,
                    activeTypes.contains(HeirType.GRANDSON), reason != null, reason,
                    names.grandsonNames.getOrNull(i).orEmpty(), activeShares, bornOutOfWedlock
                )
            }
        }
        if (input.granddaughterCount > 0) {
            val reason = blockedMap[HeirType.GRANDDAUGHTER]?.firstOrNull()?.reason
            repeat(input.granddaughterCount) { i ->
                nodes += node(
                    "granddaughter_$i", "son_0", 2, HeirType.GRANDDAUGHTER, "faraidh_heir_granddaughter", 1,
                    activeTypes.contains(HeirType.GRANDDAUGHTER), reason != null, reason,
                    names.granddaughterNames.getOrNull(i).orEmpty(), activeShares, bornOutOfWedlock
                )
            }
        }
    }

    private fun addSiblingBranch(
        nodes: MutableList<SilsilahNode>,
        input: HeirInput,
        activeTypes: Set<HeirType>,
        blockedMap: Map<HeirType, List<BlockedHeir>>,
        names: FaraidhParticipantNames,
        activeShares: List<HeirShare>,
        bornOutOfWedlock: Boolean
    ) {
        addIndividualSiblings(nodes, input.fullBrotherCount, names.fullBrotherNames, "full_bro", HeirType.FULL_BROTHER, activeTypes, blockedMap, activeShares, bornOutOfWedlock)
        addIndividualSiblings(nodes, input.fullSisterCount, names.fullSisterNames, "full_sis", HeirType.FULL_SISTER, activeTypes, blockedMap, activeShares, bornOutOfWedlock)
        addIndividualSiblings(nodes, input.paternalBrotherCount, names.paternalBrotherNames, "paternal_bro", HeirType.PATERNAL_BROTHER, activeTypes, blockedMap, activeShares, bornOutOfWedlock)
        addIndividualSiblings(nodes, input.paternalSisterCount, names.paternalSisterNames, "paternal_sis", HeirType.PATERNAL_SISTER, activeTypes, blockedMap, activeShares, bornOutOfWedlock)
        repeat(input.maternalBrotherCount) { i ->
            val reason = blockedMap[HeirType.MATERNAL_SIBLING]?.firstOrNull()?.reason
            nodes += node(
                "maternal_bro_$i", DECEASED_ID, 0, HeirType.MATERNAL_SIBLING, "faraidh_heir_maternal_sibling", 1,
                activeTypes.contains(HeirType.MATERNAL_SIBLING), reason != null, reason,
                names.maternalBrotherNames.getOrNull(i).orEmpty(), activeShares, bornOutOfWedlock
            )
        }
        repeat(input.maternalSisterCount) { i ->
            val reason = blockedMap[HeirType.MATERNAL_SIBLING]?.firstOrNull()?.reason
            nodes += node(
                "maternal_sis_$i", DECEASED_ID, 0, HeirType.MATERNAL_SIBLING, "faraidh_heir_maternal_sibling", 1,
                activeTypes.contains(HeirType.MATERNAL_SIBLING), reason != null, reason,
                names.maternalSisterNames.getOrNull(i).orEmpty(), activeShares, bornOutOfWedlock
            )
        }
    }

    private fun addIndividualSiblings(
        nodes: MutableList<SilsilahNode>,
        count: Int,
        nameList: List<String>,
        idPrefix: String,
        type: HeirType,
        activeTypes: Set<HeirType>,
        blockedMap: Map<HeirType, List<BlockedHeir>>,
        activeShares: List<HeirShare>,
        bornOutOfWedlock: Boolean
    ) {
        if (count <= 0) return
        val reason = blockedMap[type]?.firstOrNull()?.reason
        repeat(count) { i ->
            nodes += node(
                "${idPrefix}_$i", DECEASED_ID, 0, type, heirLabelKey(type), 1,
                activeTypes.contains(type), reason != null, reason,
                nameList.getOrNull(i).orEmpty(), activeShares, bornOutOfWedlock
            )
        }
    }

    private fun heirLabelKey(type: HeirType): String = when (type) {
        HeirType.HUSBAND -> "faraidh_heir_husband"
        HeirType.WIFE -> "faraidh_heir_wife"
        HeirType.FATHER -> "faraidh_heir_father"
        HeirType.MOTHER -> "faraidh_heir_mother"
        HeirType.SON -> "faraidh_heir_son"
        HeirType.DAUGHTER -> "faraidh_heir_daughter"
        HeirType.GRANDSON -> "faraidh_heir_grandson"
        HeirType.GRANDDAUGHTER -> "faraidh_heir_granddaughter"
        HeirType.FULL_BROTHER -> "faraidh_heir_full_brother"
        HeirType.FULL_SISTER -> "faraidh_heir_full_sister"
        HeirType.PATERNAL_BROTHER -> "faraidh_heir_paternal_brother"
        HeirType.PATERNAL_SISTER -> "faraidh_heir_paternal_sister"
        HeirType.MATERNAL_SIBLING -> "faraidh_heir_maternal_sibling"
        HeirType.STEP_CHILD -> "faraidh_heir_step_child"
        HeirType.UNBORN_FETUS -> "faraidh_heir_unborn_fetus"
    }

    private fun node(
        id: String,
        parentId: String,
        generation: Int,
        type: HeirType,
        labelKey: String,
        heads: Int,
        inherits: Boolean,
        blocked: Boolean,
        blockReason: BlockingReasonKey?,
        displayName: String = "",
        activeShares: List<HeirShare> = emptyList(),
        bornOutOfWedlock: Boolean = false
    ): SilsilahNode {
        val actualType = if (bornOutOfWedlock && (type == HeirType.FULL_BROTHER || type == HeirType.FULL_SISTER)) {
            HeirType.MATERNAL_SIBLING
        } else {
            type
        }
        val matchingShare = activeShares.find { it.type == actualType }
        val actualInherits = if (bornOutOfWedlock && (type == HeirType.FULL_BROTHER || type == HeirType.FULL_SISTER)) {
            activeShares.any { it.type == HeirType.MATERNAL_SIBLING }
        } else {
            inherits
        }
        val (frac, pct, amt) = if (matchingShare != null && actualInherits) {
            val indivFrac = matchingShare.fraction.divideAmongHeads(matchingShare.headCount)
            val indivPercent = matchingShare.percentage.divide(BigDecimal(matchingShare.headCount), 1, RoundingMode.HALF_UP)
            val indivAmount = matchingShare.cashAmount.divide(BigDecimal(matchingShare.headCount), 2, RoundingMode.HALF_UP)
            Triple(indivFrac.toDisplayString(), indivPercent.stripTrailingZeros().toPlainString() + "%", indivAmount)
        } else {
            Triple(null, null, null)
        }
        return SilsilahNode(
            id, parentId, generation, type, labelKey, heads, actualInherits, blocked, blockReason, displayName,
            shareFraction = frac,
            sharePercentage = pct,
            shareAmount = amt
        )
    }
}

File: iosApp/Saat/Domain/Models/FaraidhSilsilahBuilder.swift
import Foundation

struct FaraidhSilsilahBuilder {
    private static let deceasedId = "deceased"
    
    static func build(
        profile: DeceasedProfile,
        input: HeirInput,
        activeShares: [HeirShare],
        blocked: [BlockedHeir],
        names: FaraidhParticipantNames = FaraidhParticipantNames()
    ) -> [SilsilahNode] {
        var nodes: [SilsilahNode] = []
        let activeTypes = Set(activeShares.map { $0.type })
        let blockedMap = Dictionary(grouping: blocked, by: { $0.type })
        
        let deceasedLabel = profile.name.isEmpty ? "" : profile.name
        nodes.append(SilsilahNode(
            id: deceasedId,
            parentNodeId: nil,
            generationLevel: 0,
            type: .son, // Deceased placeholder
            labelKey: "faraidh_node_deceased",
            headCount: 1,
            inherits: false,
            blocked: false,
            blockReason: nil,
            displayName: deceasedLabel
        ))
        
        if input.fatherCount > 0 {
            let reason = blockedMap[.father]?.first?.reason
            nodes.append(node(
                id: "father", parentId: deceasedId, generation: -1, type: .father, labelKey: "faraidh_heir_father", heads: 1,
                inherits: activeTypes.contains(.father), blocked: reason != nil, blockReason: reason, displayName: names.fatherName, activeShares: activeShares, bornOutOfWedlock: profile.bornOutOfWedlock
            ))
        }
        
        if input.grandfatherCount > 0 {
            let reason = blockedMap[.grandfather]?.first?.reason
            let parentId = input.fatherCount > 0 ? "father" : deceasedId
            nodes.append(node(
                id: "grandfather", parentId: parentId, generation: -2, type: .grandfather, labelKey: "faraidh_heir_grandfather", heads: 1,
                inherits: activeTypes.contains(.grandfather), blocked: reason != nil, blockReason: reason, displayName: names.grandfatherName, activeShares: activeShares, bornOutOfWedlock: profile.bornOutOfWedlock
            ))
        }
        
        if input.motherCount > 0 {
            let reason = blockedMap[.mother]?.first?.reason
            nodes.append(node(
                id: "mother", parentId: deceasedId, generation: -1, type: .mother, labelKey: "faraidh_heir_mother", heads: 1,
                inherits: activeTypes.contains(.mother), blocked: reason != nil, blockReason: reason, displayName: names.motherName, activeShares: activeShares, bornOutOfWedlock: profile.bornOutOfWedlock
            ))
        }
        
        if profile.gender == .male && input.wifeCount > 0 {
            for i in 0..<input.wifeCount {
                let wifeName = names.wifeNames.indices.contains(i) ? names.wifeNames[i] : ""
                let reason = blockedMap[.wife]?.first?.reason
                nodes.append(node(
                    id: "wife_\(i)", parentId: deceasedId, generation: 0, type: .wife, labelKey: "faraidh_heir_wife",
                    heads: 1, inherits: activeTypes.contains(.wife), blocked: reason != nil, blockReason: reason, displayName: wifeName, activeShares: activeShares, bornOutOfWedlock: profile.bornOutOfWedlock
                ))
            }
        }
        
        if profile.gender == .female && input.husbandCount > 0 {
            let reason = blockedMap[.husband]?.first?.reason
            nodes.append(node(
                id: "husband", parentId: deceasedId, generation: 0, type: .husband, labelKey: "faraidh_heir_husband",
                heads: 1, inherits: activeTypes.contains(.husband), blocked: reason != nil, blockReason: reason, displayName: names.husbandName, activeShares: activeShares, bornOutOfWedlock: profile.bornOutOfWedlock
            ))
        }
        
        addChildBranch(&nodes, input: input, activeTypes: activeTypes, blockedMap: blockedMap, names: names, activeShares: activeShares, bornOutOfWedlock: profile.bornOutOfWedlock)
        addSiblingBranch(&nodes, input: input, activeTypes: activeTypes, blockedMap: blockedMap, names: names, activeShares: activeShares, bornOutOfWedlock: profile.bornOutOfWedlock)
        
        return nodes
    }
    
    private static func addChildBranch(
        _ nodes: inout [SilsilahNode],
        input: HeirInput,
        activeTypes: Set<HeirType>,
        blockedMap: [HeirType: [BlockedHeir]],
        names: FaraidhParticipantNames,
        activeShares: [HeirShare],
        bornOutOfWedlock: Bool
    ) {
        if input.sonCount > 0 {
            for i in 0..<input.sonCount {
                let reason = blockedMap[.son]?.first?.reason
                nodes.append(node(
                    id: "son_\(i)", parentId: deceasedId, generation: 1, type: .son, labelKey: "faraidh_heir_son", heads: 1,
                    inherits: activeTypes.contains(.son), blocked: reason != nil, blockReason: reason,
                    displayName: names.sonNames.indices.contains(i) ? names.sonNames[i] : "", activeShares: activeShares, bornOutOfWedlock: bornOutOfWedlock
                ))
            }
        }
        
        if input.daughterCount > 0 {
            for i in 0..<input.daughterCount {
                let reason = blockedMap[.daughter]?.first?.reason
                nodes.append(node(
                    id: "daughter_\(i)", parentId: deceasedId, generation: 1, type: .daughter, labelKey: "faraidh_heir_daughter", heads: 1,
                    inherits: activeTypes.contains(.daughter), blocked: reason != nil, blockReason: reason,
                    displayName: names.daughterNames.indices.contains(i) ? names.daughterNames[i] : "", activeShares: activeShares, bornOutOfWedlock: bornOutOfWedlock
                ))
            }
        }
        
        if input.grandsonCount > 0 {
            let reason = blockedMap[.grandson]?.first?.reason
            for i in 0..<input.grandsonCount {
                nodes.append(node(
                    id: "grandson_\(i)", parentId: "son_0", generation: 2, type: .grandson, labelKey: "faraidh_heir_grandson", heads: 1,
                    inherits: activeTypes.contains(.grandson), blocked: reason != nil, blockReason: reason,
                    displayName: names.grandsonNames.indices.contains(i) ? names.grandsonNames[i] : "", activeShares: activeShares, bornOutOfWedlock: bornOutOfWedlock
                ))
            }
        }
        
        if input.granddaughterCount > 0 {
            let reason = blockedMap[.granddaughter]?.first?.reason
            for i in 0..<input.granddaughterCount {
                nodes.append(node(
                    id: "granddaughter_\(i)", parentId: "son_0", generation: 2, type: .granddaughter, labelKey: "faraidh_heir_granddaughter", heads: 1,
                    inherits: activeTypes.contains(.granddaughter), blocked: reason != nil, blockReason: reason,
                    displayName: names.granddaughterNames.indices.contains(i) ? names.granddaughterNames[i] : "", activeShares: activeShares, bornOutOfWedlock: bornOutOfWedlock
                ))
            }
        }
    }
    
    private static func addSiblingBranch(
        _ nodes: inout [SilsilahNode],
        input: HeirInput,
        activeTypes: Set<HeirType>,
        blockedMap: [HeirType: [BlockedHeir]],
        names: FaraidhParticipantNames,
        activeShares: [HeirShare],
        bornOutOfWedlock: Bool
    ) {
        addIndividualSiblings(&nodes, count: input.fullBrotherCount, nameList: names.fullBrotherNames, idPrefix: "full_bro", type: .fullBrother, activeTypes: activeTypes, blockedMap: blockedMap, activeShares: activeShares, bornOutOfWedlock: bornOutOfWedlock)
        addIndividualSiblings(&nodes, count: input.fullSisterCount, nameList: names.fullSisterNames, idPrefix: "full_sis", type: .fullSister, activeTypes: activeTypes, blockedMap: blockedMap, activeShares: activeShares, bornOutOfWedlock: bornOutOfWedlock)
        addIndividualSiblings(&nodes, count: input.paternalBrotherCount, nameList: names.paternalBrotherNames, idPrefix: "paternal_bro", type: .paternalBrother, activeTypes: activeTypes, blockedMap: blockedMap, activeShares: activeShares, bornOutOfWedlock: bornOutOfWedlock)
        addIndividualSiblings(&nodes, count: input.paternalSisterCount, nameList: names.paternalSisterNames, idPrefix: "paternal_sis", type: .paternalSister, activeTypes: activeTypes, blockedMap: blockedMap, activeShares: activeShares, bornOutOfWedlock: bornOutOfWedlock)
        
        for i in 0..<input.maternalBrotherCount {
            let reason = blockedMap[.maternalSibling]?.first?.reason
            nodes.append(node(
                id: "maternal_bro_\(i)", parentId: deceasedId, generation: 0, type: .maternalSibling, labelKey: "faraidh_heir_maternal_sibling", heads: 1,
                inherits: activeTypes.contains(.maternalSibling), blocked: reason != nil, blockReason: reason,
                displayName: names.maternalBrotherNames.indices.contains(i) ? names.maternalBrotherNames[i] : "", activeShares: activeShares, bornOutOfWedlock: bornOutOfWedlock
            ))
        }
        
        for i in 0..<input.maternalSisterCount {
            let reason = blockedMap[.maternalSibling]?.first?.reason
            nodes.append(node(
                id: "maternal_sis_\(i)", parentId: deceasedId, generation: 0, type: .maternalSibling, labelKey: "faraidh_heir_maternal_sibling", heads: 1,
                inherits: activeTypes.contains(.maternalSibling), blocked: reason != nil, blockReason: reason,
                displayName: names.maternalSisterNames.indices.contains(i) ? names.maternalSisterNames[i] : "", activeShares: activeShares, bornOutOfWedlock: bornOutOfWedlock
            ))
        }
    }
    
    private static func addIndividualSiblings(
        _ nodes: inout [SilsilahNode],
        count: Int,
        nameList: [String],
        idPrefix: String,
        type: HeirType,
        activeTypes: Set<HeirType>,
        blockedMap: [HeirType: [BlockedHeir]],
        activeShares: [HeirShare],
        bornOutOfWedlock: Bool
    ) {
        if count <= 0 { return }
        let reason = blockedMap[type]?.first?.reason
        for i in 0..<count {
            nodes.append(node(
                id: "\(idPrefix)_\(i)", parentId: deceasedId, generation: 0, type: type, labelKey: heirLabelKey(type), heads: 1,
                inherits: activeTypes.contains(type), blocked: reason != nil, blockReason: reason,
                displayName: nameList.indices.contains(i) ? nameList[i] : "", activeShares: activeShares, bornOutOfWedlock: bornOutOfWedlock
            ))
        }
    }
    
    private static func heirLabelKey(_ type: HeirType) -> String {
        switch type {
        case .husband: return "faraidh_heir_husband"
        case .wife: return "faraidh_heir_wife"
        case .father: return "faraidh_heir_father"
        case .grandfather: return "faraidh_heir_grandfather"
        case .mother: return "faraidh_heir_mother"
        case .son: return "faraidh_heir_son"
        case .daughter: return "faraidh_heir_daughter"
        case .grandson: return "faraidh_heir_grandson"
        case .granddaughter: return "faraidh_heir_granddaughter"
        case .fullBrother: return "faraidh_heir_full_brother"
        case .fullSister: return "faraidh_heir_full_sister"
        case .paternalBrother: return "faraidh_heir_paternal_brother"
        case .paternalSister: return "faraidh_heir_paternal_sister"
        case .maternalSibling: return "faraidh_heir_maternal_sibling"
        case .stepChild: return "faraidh_heir_step_child"
        case .unbornFetus: return "faraidh_heir_unborn_fetus"
        }
    }
    
    private static func node(
        id: String,
        parentId: String,
        generation: Int,
        type: HeirType,
        labelKey: String,
        heads: Int,
        inherits: Bool,
        blocked: Bool,
        blockReason: BlockingReasonKey?,
        displayName: String = "",
        activeShares: [HeirShare] = [],
        bornOutOfWedlock: Bool = false
    ) -> SilsilahNode {
        let actualType = (bornOutOfWedlock && (type == .fullBrother || type == .fullSister)) ? HeirType.maternalSibling : type
        let matchingShare = activeShares.first { $0.type == actualType }
        let actualInherits = (bornOutOfWedlock && (type == .fullBrother || type == .fullSister)) ?
            activeShares.contains { $0.type == .maternalSibling } : inherits
            
        var frac: String? = nil
        var pct: String? = nil
        var amt: Decimal? = nil
        
        if let share = matchingShare, actualInherits {
            let indivFrac = share.fraction.divideAmongHeads(share.headCount)
            let indivPercent = share.percentage / Decimal(share.headCount)
            let indivAmount = share.cashAmount / Decimal(share.headCount)
            frac = indivFrac.toDisplayString()
            pct = "\(indivPercent)%"
            amt = indivAmount
        }
        
        return SilsilahNode(
            id: id,
            parentNodeId: parentId,
            generationLevel: generation,
            type: type,
            labelKey: labelKey,
            headCount: heads,
            inherits: actualInherits,
            blocked: blocked,
            blockReason: blockReason,
            displayName: displayName,
            shareFraction: frac,
            sharePercentage: pct,
            shareAmount: amt
        )
    }
}
